### ISSUE TITLE
Custom Execution Policies now generate warnings when calling thrust::sort / thrust::sort_by_key

### ISSUE THREAD

**Issue by robertmaynard** [thrust/thrust#992]
The following code didn't generate any warnings with thrust v1.9.5 and below, but with thrust maser I get the following warnings:

warnings:
```
/home/robert/Work/sandbox/src/ThrustExecutionPolicyWarnings/main.cu(37): warning: calling a __host__ function from a __host__ __device__ function is not allowed
          detected during instantiation of "void sort(const vtkm_cuda_policy &, T *, T *, vtkm::SortLess) [with T=int64_t]" 
(58): here

/home/robert/Work/sandbox/src/ThrustExecutionPolicyWarnings/main.cu(49): warning: calling a __host__ function from a __host__ __device__ function is not allowed
          detected during instantiation of "void sort_by_key(const vtkm_cuda_policy &, T *, T *, RandomAccessIterator, vtkm::SortLess) [with T=int64_t, RandomAccessIterator=float *]" 
(64): here
```

code:
```cpp

#include <thrust/execution_policy.h>
#include <thrust/system/cuda/execution_policy.h>
#include <thrust/system/cuda/memory.h>
#include <thrust/sort.h>

#define ThrustCudaPolicyPerThread ::thrust::cuda::par.on(cudaStreamPerThread)

namespace vtkm
{
struct SortLess
{
  template <typename T>
  __host__ __device__ bool operator()(const T& x, const T& y) const
  {
    return x < y;
  }
};
}
struct vtkm_cuda_policy : thrust::device_execution_policy<vtkm_cuda_policy>
{
};

//Specialize the sort call for cuda pointers using less/greater operators.
//The purpose of this is that for 32bit types (UInt32,Int32,Float32) thrust
//will call a super fast radix sort only if the operator is thrust::less
//or thrust::greater.
template <typename T>
__host__ __device__ void sort(
  const vtkm_cuda_policy& exec,
  T* first,
  T* last,
  vtkm::SortLess comp)
{ //sort for concrete pointers and less than op
  //this makes sure that we invoke the thrust radix sort and not merge sort
  return thrust::sort(ThrustCudaPolicyPerThread, first, last, thrust::less<T>());
}

template <typename T, typename RandomAccessIterator>
__host__ __device__ void sort_by_key(
  const vtkm_cuda_policy& exec,
  T* first,
  T* last,
  RandomAccessIterator values_first,
  vtkm::SortLess comp)
{ //sort for concrete pointers and less than op
  //this makes sure that we invoke the thrust radix sort and not merge sort
  return thrust::sort_by_key(
    ThrustCudaPolicyPerThread, first, last, values_first, thrust::less<T>());
}

void compute()
{
  vtkm_cuda_policy policy;
  {
    std::int64_t *begin = nullptr, *end = nullptr;
    sort(policy, begin, end, vtkm::SortLess{});
  }

  {
    std::int64_t *begin = nullptr, *end = nullptr;
    float *keys_begin = nullptr;
    sort_by_key(policy, begin, end, keys_begin, vtkm::SortLess{});
  }
};



```

**Comment by griwes** [comment]
The function it is being annoying about is `thrust::cuda::par.on`; there weren't any changes to how that function behaves recently. It is only marked `__device__` when device-side launches can work, i.e. with SM35 and above, and `-rdc=true`. I can't reproduce the warning with `-rdc=true`.

**Comment by robertmaynard** [comment]
Okay I am going to close the issue, as you identified the problem. I was building for SM architectures that don't support device side launches. So this warning has already existed but we avoided in the best as we normally don't build for the older SM versions.

**Comment by griwes** [comment]
Btw., here's the logic we use to determine this: https://github.com/thrust/thrust/blob/master/thrust/system/cuda/config.h#L33-L44 (...and now I'll need to do an audit of whether the `__forceinline__` there isn't overeager, _sigh_).
